### PR TITLE
feat(config): retryPredicate

### PR DESCRIPTION
**Pull request by CAMOBAP**
An alternative implementation of #73 

Previous one #85 based on `HCaptchaException.retry`. This one provide additional callback which will be fired before `onFailure()`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,7 +115,7 @@ hCaptcha.setup().verifyWithHCaptcha()
 ### Good to know
 1. The listeners (`onSuccess`, `onFailure`, `onOpen`) can be called multiple times in the following cases:
    1. the same client is used to invoke multiple verifications
-   2. the config option `resetOnTimeout(true)` is used which will automatically trigger a new verification when the current token expired. This will result in a new success or error callback.
+   2. the config option `resetOnTimeout(true)` is used which will automatically trigger a new verification when the current token expired. This will result in a new success or error callback. Also you can [retry failed challenges with `HCaptchaException.retry`](#retry-failed-challenge)
```
Comment (DSergiu) [inline README.md:118]: This is not valid.

Comment (CAMOBAP) [reply README.md]: Updated

File: example-app/src/main/java/com/hcaptcha/example/MainActivity.java [hunk 2]
```diff
@@ -68,6 +68,12 @@ private HCaptchaConfig getConfig() {
                 .hideDialog(hideDialog.isChecked())
                 .tokenExpiration(10)
                 .diagnosticLog(true)
+                .retryPredicate(new IHCaptchaRetryPredicate() {
+                    @Override
+                    public boolean shouldRetry(HCaptchaConfig config, HCaptchaException exception) {
+                        return exception.getHCaptchaError() == HCaptchaError.SESSION_TIMEOUT;
+                    }
+                })
```
Comment (CAMOBAP) [inline example-app/src/main/java/com/hcaptcha/example/MainActivity.java:76]: Should we replace it with lambda, like this?
```suggestion
                .retryPredicate((config, exception) -> exception.getHCaptchaError() == HCaptchaError.SESSION_TIMEOUT)
```

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java [hunk 3]
```diff
@@ -110,10 +110,26 @@ public class HCaptchaConfig implements Serializable {
...
+    @Builder.Default
+    private IHCaptchaRetryPredicate retryPredicate = new IHCaptchaRetryPredicate() {
+        @Override
+        public boolean shouldRetry(HCaptchaConfig config, com.hcaptcha.sdk.HCaptchaException exception) {
+            if (Boolean.TRUE.equals(config.resetOnTimeout)) {
+                return exception.getHCaptchaError() == HCaptchaError.SESSION_TIMEOUT;
+            }
+            return false;
+        }
+    };
```
Comment (CAMOBAP) [inline sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java:131]: We can still keep here a lambda instead an anonymous class declaration

```suggestion
    private IHCaptchaRetryPredicate retryPredicate = (HCaptchaConfig config, HCaptchaException exception) -> {
        if (Boolean.TRUE.equals(config.resetOnTimeout)) {
            return exception.getHCaptchaError() == HCaptchaError.SESSION_TIMEOUT;
        }
        return false;
    };
```

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaHeadlessWebView.java [hunk 4]
```diff
@@ -46,22 +46,17 @@ final class HCaptchaHeadlessWebView implements IHCaptchaVerifier {
...
-        final boolean silentRetry = webViewHelper.getConfig().getResetOnTimeout()
-                && exception.getHCaptchaError() == HCaptchaError.SESSION_TIMEOUT;
+    public void onFailure(@NonNull final HCaptchaException exception) {
+        final boolean silentRetry = webViewHelper.shouldRetry(exception);
         if (silentRetry) {
-            resetAndExecute();
+            webViewHelper.resetAndExecute();
         } else {
             listener.onFailure(exception);
         }
```
Comment (CAMOBAP) [inline sdk/src/main/java/com/hcaptcha/sdk/HCaptchaHeadlessWebView.java:62]: This block looks like the exact duplicate of https://github.com/hCaptcha/hcaptcha-android-sdk/pull/87/files#diff-d0637e553a517f03703d358d658b90ab0a930bb4c81793a93520fc074765a38dR205-R209

Previously there was `HCaptchaWebViewHelper.retryIfRequested` method, the name is probably can be better but it kept this code in a single place

Comment (DSergiu) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaHeadlessWebView.java:62]: This very simple code is not worth abstracting. The Headless and Dialog do not necessarily have the same flow.

File: example-app/src/main/java/com/hcaptcha/example/MainActivity.java (hunk 2 above)
Comment (DSergiu) [reply example-app/src/main/java/com/hcaptcha/example/MainActivity.java]: Pushed https://github.com/hCaptcha/hcaptcha-android-sdk/pull/87/commits/7a5522169c975406a2a84c63b754b284fc348602

File: sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java (hunk 3 above)
Comment (DSergiu) [reply sdk/src/main/java/com/hcaptcha/sdk/HCaptchaConfig.java]: Pushed https://github.com/hCaptcha/hcaptcha-android-sdk/pull/87/commits/7a5522169c975406a2a84c63b754b284fc348602

**Review by DSergiu — APPROVED** [review]

**Pull request merged**